### ISSUE TITLE
Hessian w.r.t. params of Neural ODE

### ISSUE THREAD

**Issue by qtomcatq** [patrick-kidger/diffrax#623]
Hello Patrick,

is there any way to compute Hessian w.r.t. parameters of Neural ODE? I tried to follow the chapter "Computing second-order sensitivities", but the output is always zero for hessian. This question is similar "Second order updates #546" in equinox depository. 

Thanks in advance 

`
import jax
import jax.numpy as jnp
import equinox as eqx
from diffrax import diffeqsolve, Tsit5, ODETerm
from jax.flatten_util import ravel_pytree

class VectorFieldMLP(eqx.Module):
    mlp: eqx.nn.MLP
    
    def __init__(self, key):
        self.mlp = eqx.nn.MLP(in_size=3, out_size=2, width_size=32, depth=5, key=key)
    
    def __call__(self, t, y):
        t_ = jnp.broadcast_to(t, (1,))
        input = jnp.concatenate([y, t_])
        return self.mlp(input)

key = jax.random.PRNGKey(67)
template_model = VectorFieldMLP(key)
t0 = 0.0
t1 = 10.0
y0 = jnp.array([10.0, 5.0])
y_target = jnp.array([0.0, 0.0])

params, unravel_fn = ravel_pytree(eqx.filter(template_model, eqx.is_inexact_array))

@jax.jit
@jax.hessian
def make_flat_loss_fn(flat_params):
    unflat_params = unravel_fn(flat_params)
    model = eqx.combine(template_model, unflat_params)
    
    def vf(t, y, args):
        return model(t, y)
    
    term = ODETerm(vf)
    solver = Tsit5(scan_kind="bounded")
    sol = diffeqsolve(term, solver, t0=t0, t1=t1, dt0=0.1, y0=y0, args=None)
    final_y = sol.ys[-1]
    return jnp.sum((final_y - y_target) ** 2)

hessian=make_flat_loss_fn(params)
print("Hessian norm:", jnp.linalg.norm(hessian, 'fro'))
`

**Comment by lockwo** [comment]
That's actually an interesting edge case of equinox. When you do combine it "leaf will be the first non-`None` leaf found in the corresponding leaves of `pytrees` as they are iterated over." and since you put the whole model in first, the parameters aren't actually being derived from the function input and thus the gradient with respect to the function input is 0. You can just reverse the order of the combine and it works or do something like

```
arr, static = eqx.partition(template_model, eqx.is_inexact_array)
params, unravel_fn = ravel_pytree(arr)
```

then just combine with the static (or just use `eqx.filter_jit`\`hessian` and feed the whole model in).

**Comment by qtomcatq** [comment]
> That's actually an interesting edge case of equinox. When you do combine it "leaf will be the first non-None leaf found in the corresponding leaves of pytrees as they are iterated over." and since you put the whole model in first, the parameters aren't actually being derived from the function input and thus the gradient with respect to the function input is 0. You can just reverse the order of the combine and it works or do something like
> 
> ```
> arr, static = eqx.partition(template_model, eqx.is_inexact_array)
> params, unravel_fn = ravel_pytree(arr)
> ```
> 
> then just combine with the static (or just use `eqx.filter_jit``hessian` and feed the whole model in).

Thanks, it seems to work in this case, and execution time is also ok for small sized MLP. Might be useful for second order methods.

**Comment by qtomcatq** [comment]
The matrices of Hessian are non-symmetric for Neural ODE. I tried different types of activation and sizes of MLP, the difference remains significantly non-zero, but it shouldn't be this way cause Cliaraut's theorem works for MLPs. Would appreciate any help with it. 


import jax
import jax.nn as jnn
import jax.numpy as jnp
import equinox as eqx
from diffrax import diffeqsolve, Tsit5, ODETerm, ReversibleHeun, Heun
from jax.flatten_util import ravel_pytree



class VectorFieldMLP(eqx.Module):
    mlp: eqx.nn.MLP

    def __init__(self, key):
        self.mlp = eqx.nn.MLP(in_size=3, out_size=2, width_size=32, depth=5, activation=jnn.swish, key=key)

    def __call__(self, t, y):
        t_ = jnp.broadcast_to(t, (1,))
        input = jnp.concatenate([y, t_])
        return self.mlp(input)


key = jax.random.PRNGKey(67)
template_model = VectorFieldMLP(key)
t0 = 0.0
t1 = 32.0
y0 = jnp.array([10.0, 5.0])
y_target = jnp.array([0.0, 0.0])

arr, static = eqx.partition(template_model, eqx.is_inexact_array)
params, unravel_fn = ravel_pytree(arr)

@eqx.filter_jit
def make_flat_loss_fn(flat_params, static):
    unflat_params = unravel_fn(flat_params)
    model = eqx.combine(static, unflat_params)

    def vf(t, y, args):
        return model(t, y)
  
    term = ODETerm(vf)
    solver = Heun(scan_kind="bounded")
    sol = diffeqsolve(term, solver, t0=t0, t1=t1, dt0=0.1, y0=y0, args=None)
    final_y = sol.ys[-1]
    return jnp.sum((final_y - y_target) ** 2)
    
hessian=jax.hessian(make_flat_loss_fn)(params, static)
print("Hessian symm diff:", jnp.linalg.norm(hessian-hessian.T, 'fro'))

**Comment by patrick-kidger** [comment]
This is likely just an accumulating of floating point error. Switching to float64 via `jax.config.update("jax_enable_x64", True)` reduces the error to `1e-11` for me.

That aside, note that you can format code (and get syntax highlighting) using backticks:
````
```python
import foo

class Bar:
    ...
```
````

**Comment by qtomcatq** [comment]
ok, thanks!
